src/handlers.rs

```diff
@@ -172,3 +172,72 @@ where
     }
     Ok(())
 }
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    macro_rules! send_to {
+        ($v:expr) => {
+            |bs| {
+                $v.extend_from_slice(bs);
+                Ok(())
+            }
+        };
+    }
+
+    fn test_logger() -> slog::Logger {
+        Logger::root(slog::Discard, slog::o!())
+    }
+
+    #[test]
+    fn test_handle_request_empty_key() {
+        let mut output = vec![];
+        let request = protocol::Request {
+            ty: protocol::RequestType::GETPWBYNAME,
+            key: &[],
+        };
+
+        assert!(
+            handle_request(&test_logger(), &request, send_to!(output)).is_err(),
+            "should error on empty input"
+        );
+        assert!(output.is_empty());
+    }
+
+    #[test]
+    fn test_handle_request_nul_data() {
+        let mut output = vec![];
+        let request = protocol::Request {
+            ty: protocol::RequestType::GETPWBYNAME,
+            key: &[0x7F, 0x0, 0x0, 0x01],
+        };
+
+        assert!(
+            handle_request(&test_logger(), &request, send_to!(output)).is_err(),
+            "should error on garbage input"
+        );
+        assert!(output.is_empty());
+    }
+
+    #[test]
+    fn test_handle_request_current_user() {
+        let current_user = User::from_uid(nix::unistd::geteuid()).unwrap().unwrap();
+
+        let request = protocol::Request {
+            ty: protocol::RequestType::GETPWBYNAME,
+            key: &CString::new(current_user.name.clone())
+                .unwrap()
+                .into_bytes_with_nul(),
+        };
+
+        let mut expected = vec![];
+        send_user(&test_logger(), Some(current_user), send_to!(expected))
+            .expect("send_user should serialize current user data");
+
+        let mut output: Vec<u8> = vec![];
+        handle_request(&test_logger(), &request, send_to!(output))
+            .expect("should handle request with no error");
+        assert_eq!(expected, output,);
```

Comment:
Extra comma?

Reply:
weird, that survived `rustfmt`. I'll remove it.